lib/libc/stdlib/malloc/mrs/mrs.c

```diff
@@ -0,0 +1,1341 @@
+
+	/*
+	 * we trigger a revocation pass when the unvalidated quarantine hits the
+	 * highwater mark because if we waited until the validated queue passed the
+	 * highwater mark, the allocated size might increase (allocations made) between
+	 * the unvalidated queue and validated queue filling such that the high water
+	 * mark is no longer hit. this function just fills up the unvalidated
+	 * quarantine and passes it off when it's full. with offload enabled,
+	 * the "quarantine" global is unvalidated and passed off to the
+	 * "offload_quarantine" global then processed in place (list entries that fail
+	 * validation are not processed).
+	 */
+	if (quarantine_should_flush(&application_quarantine)) {
+#ifdef PRINT_CAPREVOKE
+		mrs_printf("check_and_perform_flush (offload): passed application_quarantine threshold, offloading: allocated size %zu quarantine size %zu\n", allocated_size, application_quarantine.size);
+#endif
+#ifdef PRINT_CAPREVOKE_MRS
+		mrs_printf("check_and_perform flush: cycle count before waiting on offload %" PRIu64 "\n", cheri_revoke_get_cyc());
+#endif
+
+		malloc_revoke();
+	}
+
+#else /* OFFLOAD_QUARANTINE */
+
+	if (quarantine_should_flush(&application_quarantine)) {
+#ifdef PRINT_CAPREVOKE
+		mrs_printf("check_and_perform_flush (no offload): passed application_quarantine threshold, revoking: allocated size %zu quarantine size %zu\n", allocated_size, application_quarantine.size);
+#endif
+		malloc_revoke();
+	}
+
+#endif /* !OFFLOAD_QUARANTINE */
+}
+
+/* constructor and destructor */
+
+#ifdef OFFLOAD_QUARANTINE
+static void spawn_background(void)
+{
+	initialize_lock(offload_quarantine_lock);
+
+	/*
+	 * Fire a spurious signal at the condvars now in an attempt to force
+	 * the pthread machinery to initialize them fully now, before the
+	 * application has a chance to start up and fill quarantine.  Otherwise
+	 * there's a chance that we'll attempt allocation while signaling the
+	 * need for revocation during allocation (thereby deadlocking an
+	 * application thread against itself) or while signaling that the
+	 * quarantine is empty (thereby deadlocking the offload thread against
+	 * itself).
+	 */
+	pthread_cond_signal(&offload_quarantine_empty);
+	pthread_cond_signal(&offload_quarantine_ready);
+
+	pthread_t thd;
+	if (pthread_create(&thd, NULL, mrs_offload_thread, NULL)) {
+		mrs_puts("pthread error\n");
+		exit(7);
+	}
+}
+#endif /* OFFLOAD_QUARANTINE */
+
+__attribute__((constructor))
+static void init(void) {
+	real_malloc = dlsym(RTLD_NEXT, "malloc");
+	real_free = dlsym(RTLD_NEXT, "free");
+	real_calloc = dlsym(RTLD_NEXT, "calloc");
+	real_realloc = dlsym(RTLD_NEXT, "realloc");
+	real_posix_memalign = dlsym(RTLD_NEXT, "posix_memalign");
+	real_aligned_alloc = dlsym(RTLD_NEXT, "aligned_alloc");
+
+	initialize_lock(printf_lock);
+
+#ifdef OFFLOAD_QUARANTINE
+	spawn_background();
+	pthread_atfork(NULL, NULL, spawn_background);
+#endif /* OFFLOAD_QUARANTINE */
+
+#ifdef MRS_PINNED_CPUSET
+#ifdef __aarch64__
```

Comment:
Why are these #ifdef'ed?

Reply:
It looks like this is a hack for Morello. We're not enabling MRS_PINNED_CPUSET.